### PR TITLE
Handle duplicate pixels in sparse pixel gaussian rendering

### LINKED ISSUE #106: Handle duplicate pixels in sparse pixel gaussian rendering
## 📋 Originally created by @fwilliams

> **Original author:** @fwilliams
> **Date:** 2025-02-12T00:03:49Z

---

We will squash duplicate pixels in the current code. Let's handle this properly.

### PR DESCRIPTION
**Pull request by harrism**
## Summary

- Deduplicate pixel coordinates at the binding layer (`deduplicatePixels`) before passing to `computeSparseInfo` and rasterization kernels, then scatter results back via `index_select`. This avoids any CUDA kernel modifications while correctly handling duplicate `(batchIdx, row, col)` entries that would otherwise cause incorrect tile bitmasks.
- Handle edge cases: single-list JaggedTensors with empty `jidx`, correct `jlidx` passthrough for `ldim()` consistency, and reducing `numContributingGaussians` to unique-pixel space for the contributing IDs kernel.
- Add comprehensive C++ unit tests for `deduplicatePixels` (20 tests covering empty, single, all-unique, duplicates, multi-batch, round-trip reconstruction, and per-batch offset validation) and Python end-to-end tests for all sparse render APIs with duplicate pixels (6 tests covering depths, images, backward gradients, num_contributing, contributing_ids, and multi-camera).

Fixes #106

## Test plan

- [x] C++ `DeduplicatePixelsTest`: 20/20 pass
- [x] C++ `GaussianComputeSparseInfoTest`: 14/14 pass
- [x] Python `TestGaussianRenderSparseDuplicatePixels`: 6/6 pass

### CODE REVIEW

File: src/fvdb/GaussianSplat3d.cpp [hunk 1]
```diff
@@ -392,6 +390,89 @@ GaussianSplat3d::projectGaussiansImpl(const torch::Tensor &worldToCameraMatrices
...
+    // Build new JaggedTensor offsets for the unique pixels
+    auto uniqueBatchIdx = jidxLong.index_select(0, uniqueOrigIndices);
+    auto numLists       = pixelsToRender.num_outer_lists();
+    auto countsPerList  = torch::zeros({numLists}, longOpts);
+    countsPerList.scatter_add_(
+        0, uniqueBatchIdx.to(torch::kLong), torch::ones_like(uniqueBatchIdx, torch::kLong));
+    auto newOffsets = torch::zeros({numLists + 1}, longOpts);
+    newOffsets.slice(0, 1).copy_(countsPerList.cumsum(0));
+
+    auto newJidx = JaggedTensor::jidx_from_joffsets(newOffsets, numUnique);
```
Comment (swahtz) [inline src/fvdb/GaussianSplat3d.cpp]: **💡 suggestion:** ‏I think `uniqueBatchIdx` already has the values you want for the new jagged indices.  You could just do this and save the binary search + allocation

```suggestion
    auto newJidx = uniqueBatchIdx.to(fvdb::JIdxScalarType);
```

File: src/fvdb/GaussianSplat3d.cpp [hunk 2]
```diff
@@ -392,6 +390,89 @@ GaussianSplat3d::projectGaussiansImpl(const torch::Tensor &worldToCameraMatrices
...
+    auto firstInSorted     = isGroupStart.nonzero().squeeze(1);
+    auto uniqueOrigIndices = sortPerm.index_select(0, firstInSorted);
+    auto uniqueJData       = jdata.index_select(0, uniqueOrigIndices);
+
+    // Build new JaggedTensor offsets for the unique pixels
+    auto uniqueBatchIdx = jidxLong.index_select(0, uniqueOrigIndices);
+    auto numLists       = pixelsToRender.num_outer_lists();
+    auto countsPerList  = torch::zeros({numLists}, longOpts);
+    countsPerList.scatter_add_(
+        0, uniqueBatchIdx.to(torch::kLong), torch::ones_like(uniqueBatchIdx, torch::kLong));
```
Comment (swahtz) [inline src/fvdb/GaussianSplat3d.cpp]: **💡 suggestion:** ‏`uniqueBatchIdx` is already a long because it came from `jidxLong` so you could save a dispatch.  `torch::ones` can also be used instead of  `ones_like` (save on querying the shape of `uniqueBatchIdx` which we already have in `numUnique`).

```suggestion
        0, uniqueBatchIdx, torch::ones({numUnique}, longOpts));
```

Comment (swahtz) [inline src/fvdb/GaussianSplat3d.cpp]: **💡 suggestion:** ‏ On further thought, we could use `torch::bincount` to do these 2 lines in 1 call instead of 3.

```suggestion
    auto countsPerList = torch::bincount(uniqueBatchIdx, {}, numLists);
```

**Review by swahtz — APPROVED** [review]
Looks great I just had a few potential optimization suggestions since these will get called on every rendering iteration. 

I was also wondering for the unit tests if it would be worth having a test where all pixels in each batch are duplicates or perhaps that's redundant with the tests were just some pixels are duplicates.

File: src/fvdb/GaussianSplat3d.cpp (hunk 2 above)
Comment (harrism) [reply src/fvdb/GaussianSplat3d.cpp]: Great catch!

Comment (harrism) [reply src/fvdb/GaussianSplat3d.cpp]: Took the other suggestion.

File: src/fvdb/GaussianSplat3d.cpp (hunk 1 above)
Comment (harrism) [reply src/fvdb/GaussianSplat3d.cpp]: Wow, this is what I get for trusting the LLM. Good catch.

**Comment by harrism** [comment]
Great suggestions @swahtz . Implemented your two optimizations and found one other and a stale comment. Also added your suggested maximum compression test case.

File: src/fvdb/GaussianSplat3d.cpp [hunk 3]
```diff
@@ -393,6 +393,87 @@ GaussianSplat3d::projectGaussiansImpl(const torch::Tensor &worldToCameraMatrices
...
+    const auto jidx                 = pixelsToRender.jidx();
+    const int64_t numPixelsPerImage = imageHeight * imageWidth;
+    const auto longOpts             = torch::TensorOptions().device(device).dtype(torch::kLong);
+    const auto boolOpts             = torch::TensorOptions().device(device).dtype(torch::kBool);
+
+    // Encode (batchIdx, row, col) into a single int64 key:
+    //   key = batchIdx * (H * W) + row * W + col
+    // For single-list JaggedTensors, jidx may be empty; synthesize zeros.
+    auto jidxLong =
+        jidx.size(0) == 0 ? torch::zeros({totalPixels}, longOpts) : jidx.to(torch::kLong);
```
Comment (fwilliams) [inline src/fvdb/GaussianSplat3d.cpp]: why not use an empty jidxLong instead of zeros here?

File: src/fvdb/GaussianSplat3d.cpp [hunk 4]
```diff
@@ -393,6 +393,87 @@ GaussianSplat3d::projectGaussiansImpl(const torch::Tensor &worldToCameraMatrices
...
+    // Sort keys and find group boundaries
+    auto [sortedKeys, sortPerm] = keys.sort();
+
+    auto isGroupStart = torch::ones({totalPixels}, boolOpts);
+    if (totalPixels > 1) {
+        isGroupStart.slice(0, 1).copy_(sortedKeys.slice(0, 1) != sortedKeys.slice(0, 0, -1));
+    }
+
+    // Assign a group ID (0-based) to each sorted position via cumsum
+    auto groupIds        = isGroupStart.cumsum(0) - 1;
```
Comment (fwilliams) [inline src/fvdb/GaussianSplat3d.cpp]: you can do an in place cumsum instead of creating a new tensor.  you'll have to compute firstInSorted before you do this (it's currently below)

File: src/fvdb/GaussianSplat3d.cpp (hunk 3 above)
Comment (harrism) [reply src/fvdb/GaussianSplat3d.cpp]: Can do this. It requires some branching but that's OK. Saves a bit of memory and fill and it's easy.

File: src/fvdb/GaussianSplat3d.cpp (hunk 4 above)
Comment (harrism) [reply src/fvdb/GaussianSplat3d.cpp]: Good point.

**Comment by harrism** [comment]
I decided against the custom kernel approach because by definition these are low-cost operations in terms of memory and computation. For sparse rendering, N (number of pixels) should be small -- say 5-50K. I estimate memory savings of a custom kernel to be at most ~1MB at that scale. Even if we render at the scale of 500K pixels (25% of a 1080p image) we're talking only 10MB savings. On the scale of 3DGS memory usage, this is peanuts.

The main reason for custom kernels here would be launch overhead and possibly synchronization reduction. But I don't think that should be addressed until it is determined to be a bottleneck.

**Comment by harrism** [comment]
@fwilliams implemented the small optimizations you suggested. Torch cumsum_ doesn't work on bool tensors in CUDA, so had to cast it, so the memory savings is reduced a bit but there's still about 8N bytes savings (1 fewer allocation) for that one (rather than 16N).
